frontend/src/metabase/lib/expressions/tokenizer.js

```diff
@@ -0,0 +1,328 @@
+        const end = index;
+        const error = t`Missing exponent`;
+        return { type, start, end, error };
+      }
+    }
+    const type = TOKEN.Number;
+    const end = index;
+    const error = null;
+    return { type, start, end, error };
+  };
+
+  const scanStringLiteral = () => {
+    const start = index;
+    const quote = source[start];
+    if (quote !== "'" && quote !== '"') {
+      return null;
+    }
+    ++index;
+    let value = "";
+    while (index < length) {
+      const ch = source[index++];
+      if (ch === quote) {
+        break;
+      } else if (ch === "\\") {
+        const seq = source[index++];
+        if (seq) {
+          switch (seq) {
+            case "b":
+              value += "\b";
+              break;
+            case "f":
+              value += "\f";
+              break;
+            case "n":
+              value += "\n";
+              break;
+            case "r":
+              value += "\r";
+              break;
+            case "t":
+              value += "\t";
+              break;
+            case "v":
+              value += "\x0b";
+              break;
+            default:
+              value += seq;
+              break;
+          }
+        }
+      } else {
+        value += ch;
+      }
+    }
+    const type = TOKEN.String;
+    const end = index;
+    const terminated = quote === source[end - 1];
+    const error = terminated ? null : t`Unterminated quoted string`;
+    return { type, value, start, end, error };
+  };
+
+  const scanBracketIdentifier = () => {
+    const start = index;
+    const bracket = source[start];
+    if (bracket !== "[") {
+      return null;
+    }
+    ++index;
+    while (index < length) {
+      const ch = source[index++];
+      if (ch === "]") {
+        break;
+      } else if (ch === "[") {
+        const type = TOKEN.Identifier;
+        const end = index;
+        const error = t`Bracket identifier in another bracket identifier`;
+        return { type, start, end, error };
+      } else if (ch === "\\") {
+        // ignore the next char, even if it's [ or ]
+        index++;
+      }
+    }
+    const type = TOKEN.Identifier;
+    const end = index;
+    const terminated = source[end - 1] === "]";
+    const error = terminated ? null : t`Unterminated bracket identifier`;
+    return { type, start, end, error };
+  };
+
+  const isIdentifierStart = cp => isAlpha(cp) || cp === 0x5f; // underscore;
+
+  const isIdentifierChar = cp =>
+    isAlpha(cp) ||
+    isDigit(cp) ||
+    cp === 0x2e || // dot
+    cp === 0x5f; // underscore
+
+  const scanIdentifier = () => {
+    const start = index;
+    const initial = source.charCodeAt(start);
+    if (!isIdentifierStart(initial)) {
+      return null;
+    }
+
+    while (index < length) {
+      const cp = source.charCodeAt(index);
+      if (!isIdentifierChar(cp)) {
+        break;
+      }
+      ++index;
+    }
+    const end = index;
+    if (index <= start) {
```

Comment:
is there a scenario here where `index` could be less than `start`? not sure I see it

Reply:
Good point! I'll adjust is accordingly.